src/WebJobs.Script/Binding/Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -0,0 +1,320 @@
+        }
+
+        internal class HttpTriggerBinding : ITriggerBinding
+        {
+            private readonly ParameterInfo _parameter;
+            private readonly IBindingDataProvider _bindingDataProvider;
+            private readonly bool _isUserTypeBinding;
+
+            public HttpTriggerBinding(ParameterInfo parameter, bool isUserTypeBinding)
+            {
+                _parameter = parameter;
+                _isUserTypeBinding = isUserTypeBinding;
+
+                if (_isUserTypeBinding)
+                {
+                    // Create the BindingDataProvider from the user Type. The BindingDataProvider
+                    // is used to define the binding parameters that the binding exposes to other
+                    // bindings (i.e. the properties of the POCO can be bound to by other bindings).
+                    // It is also used to extract the binding data from an instance of the Type.
+                    _bindingDataProvider = BindingDataProvider.FromType(parameter.ParameterType);
+                }
+            }
+
+            public IReadOnlyDictionary<string, Type> BindingDataContract
+            {
+                get
+                {
+                    // if we're binding to a user Type, we'll have a contract,
+                    // otherwise none
+                    return _bindingDataProvider != null ? _bindingDataProvider.Contract : null;
+                }
+            }
+
+            public Type TriggerValueType
+            {
+                get { return typeof(HttpRequestMessage); }
+            }
+
+            public async Task<ITriggerData> BindAsync(object value, ValueBindingContext context)
+            {
+                HttpRequestMessage request = value as HttpRequestMessage;
+                if (request == null)
+                {
+                    throw new NotSupportedException("An HttpRequestMessage is required");
+                }
+
+                IValueProvider valueProvider = null;
+                IReadOnlyDictionary<string, object> bindingData = null;
+                string invokeString = request.ToString();
+                if (_isUserTypeBinding)
+                {
+                    valueProvider = await CreateUserTypeValueProvider(request, invokeString);
+                    if (_bindingDataProvider != null)
+                    {
+                        // binding data is defined by the user type
+                        // the provider might be null if the Type is invalid, or if the Type
+                        // has no public properties to bind to
+                        bindingData = _bindingDataProvider.GetBindingData(valueProvider.GetValue());
+                    }
+                }
+                else
+                {
+                    valueProvider = new HttpRequestValueBinder(_parameter, request, invokeString);
+                    bindingData = GetRequestBindingData(request);
```

Comment:
I actually started with the merged approach you mentioned, but moved away from it since I couldn't think of strong scenarios that required it. If you're binding to a POCO and accepting a POST body, you can easily send the additional values along in the body rather than the URI. I think either/or maps to what people will actually do in practice. Also, you start having to deal with precedence (same property sent in body and URI), etc.

What do others think?